src/main/scala/com/fulcrumgenomics/alignment/Matrix.scala

```diff
@@ -0,0 +1,101 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.alignment
+
+import com.fulcrumgenomics.FgBioDef._
+
+import scala.reflect.ClassTag
+
+/** Factory method(s) for Matrices. */
+object Matrix {
+  def apply[A: ClassTag](x: Int, y: Int): Matrix[A] = {
+    require(x >= 0, s"X dimension of matrix ($x) is less than 0.")
+    require(y >= 0, s"Y dimension of matrix ($y) is less than 0.")
+    new LinearMatrix(x, y)
+  }
+}
+
+
+/**
+  * Defines methods applicable to a 2D matrix.  Matrix cells are accessed with a pair of
+  * coordiantes, the first of which is the row offset starting at 0, and the second of
+  * which is the column offset, starting at 0.
+  * */
+trait Matrix[A] {
+  /** Access a cell in the matrix. */
+  def apply(i: Int, j: Int): A
+
+  /** Updates a cell in the matrix. */
+  def update(i: Int, j: Int, value: A): Unit
+
+  /**
+    * Returns the dimensions of the matrix.  E.g. a 2x2 matrix will return (2,2), although the maximum
+    * valid offsets are (1,1).
```

Comment:
Meaning in a 2x2 matrix, the dimensions are (2,2), but the last cell is accessed at `(1,1)` because it's all zero-based like arrays.  I'll try again.